### PR TITLE
GEOMESA-601,661 Resolution Picking and Mosaic Scaling/Cropping

### PR DESCRIPTION
**Pull request by aannex-ccri**
* QueryPlanner now selects resolution for wms viewing.
* Mosaic code now crops/resamples chunks as needed.
* Added dependency imgscalr.
* Added more mosaic tests.

### CODE REVIEW

File: geomesa-plugin/src/main/scala/org/locationtech/geomesa/plugin/wcs/GeoMesaCoverageReader.scala [hunk 1]
```diff
@@ -79,7 +79,11 @@ class GeoMesaCoverageReader(val url: String, hints: Hints) extends AbstractGridC
     logger.debug(s"READ: $parameters")
     val params = new GeoMesaCoverageQueryParams(parameters)
     val rq = params.toRasterQuery
-    this.coverageFactory.create(coverageName, ars.getMosaicedRaster(rq, params), params.envelope)
+    println(s"In rastersToCoverage: width: ${params.width.toInt} height: ${params.height.toInt} resX: ${params.resX} resY: ${params.resY} env: ${params.envelope}")
```
Comment (jnh5y) [inline geomesa-plugin/src/main/scala/org/locationtech/geomesa/plugin/wcs/GeoMesaCoverageReader.scala]: I know I put in the println, but we should change it to use logger.

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala [hunk 2]
```diff
@@ -83,46 +107,110 @@ object RasterUtils {
...
-        count += 1
+      val firstRaster = chunks.next()
+      val accumuloRasterXRes = firstRaster.referencedEnvelope.getSpan(0) / firstRaster.chunk.getWidth
+      val accumuloRasterYRes = firstRaster.referencedEnvelope.getSpan(1) / firstRaster.chunk.getHeight
+      val mosaicX = (queryEnv.getSpan(0) / accumuloRasterXRes).toInt
+      val mosaicY = (queryEnv.getSpan(1) / accumuloRasterYRes).toInt
+      if (mosaicX <= 0 || mosaicY <= 0) {
+        (getEmptyImage(1, 1, BufferedImage.TYPE_BYTE_GRAY), count)
+      } else {
+        val mosaic = allocateBufferedImage(mosaicX, mosaicY, firstRaster.chunk)
```
Comment (jnh5y) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala]: @aannex-ccri Can you add a ticket for putting in a check to see if we are only handling one image?  If we are, then we can save a little time mosaicing.

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala [hunk 3]
```diff
@@ -145,7 +233,7 @@ object RasterUtils {
   }
 
   def createRasterStore(tableName: String) = {
-    val rs = RasterStore("user", "pass", "testInstance", "zk", tableName, "SUSA", "SUSA", true)
+    val rs = RasterStore("user", "pass", "testInstance", "zk", tableName, "", "", true)
```
Comment (jnh5y) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala]: How is this function used?  We can deal with it during some clean-up PRs.

File: geomesa-raster/src/test/scala/org/locationtech/geomesa/raster/data/RasterStoreQueryIntegratedTest.scala [hunk 4]
```diff
@@ -169,21 +169,21 @@ class RasterStoreQueryIntegratedTest extends Specification {
...
-      rasterStore.putRaster(testRaster)
-
-      //generate query
-      val query = RasterUtils.generateQuery(0, 50, 0, 50, res = 10.0)
-
-      rasterStore must beAnInstanceOf[RasterStore]
-      val theResults = rasterStore.getRasters(query).toList
-      theResults.length must beEqualTo(0)
-    }
+//    "Properly filter out a raster conforming to GeoHashes via a query bbox and resolution" in {
```
Comment (jnh5y) [inline geomesa-raster/src/test/scala/org/locationtech/geomesa/raster/data/RasterStoreQueryIntegratedTest.scala]: I probably did this.  Does the test work again?  

If not, when committing, I prefer 'pendingUntilFixed'.

Comment (aannex-ccri) [inline geomesa-raster/src/test/scala/org/locationtech/geomesa/raster/data/RasterStoreQueryIntegratedTest.scala]: this test is probably not needed anymore as we now select the resolution from what is available in the table, this test originally showed that results could be filtered by resolution.

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala (hunk 3 above)
Comment (aannex-ccri) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala]: this is just creating a mock table for the various unit tests

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala (hunk 2 above)
Comment (aannex-ccri) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala]: I think we currently handle this case in a reasonable way, we always mosaic the first image in the iterator and then only do further work if the iterator hasNext.

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala (hunk 3 above)
Comment (jnh5y) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala]: I'd move this out into a test class/object, something like RasterTestUtils should be good.

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala (hunk 2 above)
Comment (jnh5y) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala]: For the one tile request case, I think we can do something a little quicker than calling allocateBufferedImage, writeToMosaic, and scaleBufferedImage.  That's my intent.

In general, I agree that this code is reasonable.  I'm suggesting more (unreasonable) handling in order to speed up specific corner cases (to be unreasonably fast;)).

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala (hunk 3 above)
Comment (aannex-ccri) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala]: done.

File: geomesa-plugin/src/main/scala/org/locationtech/geomesa/plugin/wcs/GeoMesaCoverageReader.scala (hunk 1 above)
Comment (aannex-ccri) [inline geomesa-plugin/src/main/scala/org/locationtech/geomesa/plugin/wcs/GeoMesaCoverageReader.scala]: done.

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala (hunk 2 above)
Comment (aannex-ccri) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala]: ok I see what you mean, I will put that in a new ticket and reference in a todo.

**Comment by jnh5y** [comment]
Committed as 9acc9e827875bf86ee9500205a7e4240ebe88f39 on f_wcs_4.

**Pull request closed without merging**